File: myExpenses/src/test/java/org/totschnig/myexpenses/sync/ChangeReaderTest.java
package org.totschnig.myexpenses.sync;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.junit.Before;
import org.junit.Test;
import org.totschnig.myexpenses.sync.json.AdapterFactory;
import org.totschnig.myexpenses.sync.json.TransactionChange;
import org.totschnig.myexpenses.sync.json.Utils;

import java.io.StringReader;
import java.util.List;
import java.util.Locale;

import static org.junit.Assert.*;

public class ChangeReaderTest {
  private Gson gson;

  @Before
  public void setup() {
    gson = new GsonBuilder()
        .registerTypeAdapterFactory(AdapterFactory.create())
        .create();
  }

  @Test
  public void shouldParseIntoListOfChanges() {
    TransactionChange.Type created = TransactionChange.Type.created;
    String uuid = "825ec542-a434-4954-b59e-e47b71138b35";
    long timestamp = 1475560175;
    long date = 1475559751;
    long amount = -12300;
    String crStatus = "UNRECONCILED";
    TransactionChange expected = TransactionChange.builder().setType(created).setUuid(uuid).setTimeStamp(timestamp).setDate(date).setAmount(amount).setCrStatus(crStatus).build();
    StringReader reader = new StringReader(String.format(Locale.US,
        "[{\"type\":\"%s\",\"uuid\":\"%s\",\"timeStamp\":%d,\"date\":%d,\"amount\":%d,\"crStatus\":\"%s\"}]",
        created, uuid, timestamp, date, amount, crStatus));
    List<TransactionChange> result = Utils.getChanges(gson, reader);
    assertEquals(1, result.size());
    assertEquals(expected, result.get(0));
  }
}

File: myExpenses/src/test/java/org/totschnig/myexpenses/sync/ChangeReaderTest.kt
package org.totschnig.myexpenses.sync

import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.totschnig.myexpenses.sync.json.TransactionChange
import org.totschnig.myexpenses.sync.json.Utils


class ChangeReaderTest {


    @Test
    fun shouldParseIntoListOfChanges() {
        val uuid = "825ec542-a434-4954-b59e-e47b71138b35"
        val timestamp: Long = 1475560175
        val date: Long = 1475559751
        val amount: Long = -12300
        val crStatus = "UNRECONCILED"
        val expected = TransactionChange(
            type = TransactionChange.Type.created,
            uuid = uuid,
            timeStamp = timestamp,
            date = date,
            amount = amount,
            crStatus = crStatus
        )

        val json = String.format(
            java.util.Locale.US,
            "[{\"type\":\"created\",\"uuid\":\"%s\",\"timeStamp\":%d,\"date\":%d,\"amount\":%d,\"crStatus\":\"%s\"}]",
            uuid, timestamp, date, amount, crStatus
        )
        val result = Utils.getChanges(json)
        assertThat(result).hasSize(1)
        assertThat(result).containsExactly(expected)
    }
}
